### PR TITLE
raise error if release label contains forbidden characters

### PR DESCRIPTION
**Pull request by rcerven**
Signed-off-by: Robert Cerven <rcerven@redhat.com>

### CODE REVIEW

**Comment by rcerven** [comment]
still no unit tests

**Comment by twaugh** [comment]
Would be nicer to match against a regular express describing a valid release label. For example " " and "-" should also be disallowed, just off the top of my head, and there are going to be plenty of others.

**Comment by rcerven** [comment]
added unit test, and matching characters against regular expression

**Comment by twaugh** [comment]
My preference would be for the regular expression to describe *valid* release labels. Something like: `^[0-9]([._]?[a-zA-Z0-9]+)*$`. Even better if it uses a verbose regexp to describe it.

**Comment by rcerven** [comment]
regex is now matching valid values

File: osbs/constants.py [hunk 1]
```diff
@@ -102,6 +102,9 @@
 BUILD_TYPE_WORKER = "worker"
 
 ISOLATED_RELEASE_FORMAT = re.compile(r'^\d+\.\d+(\..+)?$')
+RELEASE_LABEL_FORMAT = re.compile(r"""^\d+             # First character must be a digit
+                                      ([._]?          # allow separators between groups
+                                      [a-zA-Z0-9]+)*$  # last characters must be alphanumeric""", re.X)
```
[the remark this answers is not in the document]
Comment (twaugh) [reply osbs/constants.py]: Could shorten line lengths:
```
RELEASE_LABEL_FORMAT = re.compile(r"""
    ^\d+ ...
""", re.X)

File: osbs/api.py [hunk 2]
```diff
@@ -491,6 +492,14 @@ def _check_labels(self, repo_info):
                 logger.error("required label missing from Dockerfile : %s",
                              labels.get_name(label))
 
+        try:
+            _, release_value = labels.get_name_and_value(utils.Labels.LABEL_TYPE_RELEASE)
+            if not RELEASE_LABEL_FORMAT.match(release_value):
+                logger.error("release label contains forbidden characters : %s", release_value)
+                raise OsbsValidationException("release label contains '/'")
```
Comment (lcarva) [inline osbs/api.py]: This error message is misleading. If the release label is `'abc'` user will get this message and be quite confused. Can we instead output the expected regex? `RELEASE_LABEL_FORMAT.pattern` might give you what you want.

File: osbs/constants.py [hunk 3]
```diff
@@ -102,6 +102,10 @@
 BUILD_TYPE_WORKER = "worker"
 
 ISOLATED_RELEASE_FORMAT = re.compile(r'^\d+\.\d+(\..+)?$')
+RELEASE_LABEL_FORMAT = re.compile(r"""^\d+             # First character must be a digit
+                                      ([._]?           # allow separators between groups
+                                      [a-zA-Z0-9]+)*$  # last characters must be alphanumeric""",
```
Comment (lcarva) [inline osbs/constants.py]: I really enjoy this way of wring regexes, much easier to read!

A minor improvement would be to move the closing `""",` to a new line.

File: osbs/api.py (hunk 2 above)
Comment (twaugh) [reply osbs/api.py]: Yes... once you remove whitespace and comments. :-)
Perhaps `re.sub('\s', '', re.sub(r' #.*(\n|$)', '', RELEASE_LABEL_FORMAT.pattern))`

Comment (rcerven) [reply osbs/api.py]: sure

File: osbs/constants.py (hunk 3 above)
Comment (rcerven) [reply osbs/constants.py]: sure

**Comment by lcarva** [comment]
@mlangsdorf, I think your requested changes have been addressed. PTAL

**Comment by rcerven** [comment]
rn updated: raise error if release label contains forbidden characters

**Pull request merged**